include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright (c) 2022 Alan Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+#define BOOST_URL_IMPL_ANY_ALLOCATOR_HPP
+
+#include <atomic>
+
+#include <boost/url/detail/over_allocator.hpp>
+
+namespace boost {
+namespace urls {
+
+// base class to store the underlying allocator bound to `base`
+template <class T>
+struct any_allocator<T>::base
+{
+  std::atomic<std::size_t> refs{1};
+
+  virtual
+  ~base()
+  {
+  }
+
+  // although the allocator in `base` is bound to `base`, `allocate` always
+  // returns `char` as make_string_value does.
+  virtual
+  char*
+  allocate(std::size_t n) = 0;
+
+  virtual
+  void
+  deallocate(char*p, std::size_t n) = 0;
+
+  template <class U>
+  void
+  deallocate(U* p, std::size_t n) {
+      deallocate(reinterpret_cast<char*>(p), n * sizeof(T));
+  }
+
+  void
+  deallocate(char*p) {
+      deallocate(p, 1);
+  }
+
+  template <class U>
+  void
+  deallocate(U* p) {
+      deallocate(p, 1);
+  }
+
+  virtual void destroy_self() noexcept = 0;
+
+};
+
+template <class T>
+any_allocator<T>::
+~any_allocator()
+{
+  if( p_ && --p_->refs == 0)
+    p_->destroy_self();
+}
+
+template <class T>
+any_allocator<T>&
+any_allocator<T>::
+operator=(
+    any_allocator<T> const& other) noexcept
+{
+  if(other.p_)
+    ++other.p_->refs;
+  if( p_ && --p_->refs == 0)
+    p_->destroy_self();
+  p_ = other.p_;
+  return *this;
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a)
+{
+  // declare a concrete implementation of `base` using the specified allocator
+  class impl;
+
+  // an allocator we can use to allocator `base`
+  using char_allocator_type = typename
+      detail::allocator_traits<
+          Allocator>::template
+      rebind_alloc<char>;
```

Comment:
I think rebinding to `max_align_t` and changing `allocate` and `deallocate` will make things aligned correctly

Reply:
actually no that's not good. we should also pass `std::size_t align` to `base::allocate`

Reply:
I'm implementing a solution to that.